File: CarHomeWidget/src/main/java/com/anod/car/home/skin/icon/CardsBackgroundProcessor.java
package com.anod.car.home.skin.icon;

import com.anod.car.home.prefs.model.WidgetSettings;

import android.graphics.Bitmap;
import android.graphics.Color;
import androidx.palette.graphics.Palette;

/**
 * @author alex
 * @date 2014-12-06
 */
public class CardsBackgroundProcessor implements BackgroundProcessor {

    @Override
    public int getColor(WidgetSettings prefs, Bitmap icon) {
        if (icon == null) {
            return Color.DKGRAY;
        }
        Palette palette = new Palette.Builder(icon).generate();

        Palette.Swatch swatch = palette.getMutedSwatch();

        if (swatch != null) {
            return swatch.getRgb();
        }
        swatch = palette.getDarkVibrantSwatch();
        if (swatch != null) {
            return swatch.getRgb();
        }
        swatch = palette.getVibrantSwatch();
        if (swatch != null) {
            return swatch.getRgb();
        }
        return Color.DKGRAY;
    }
}

File: CarHomeWidget/src/main/java/com/anod/car/home/skin/icon/CardsBackgroundProcessor.kt
package com.anod.car.home.skin.icon

import com.anod.car.home.prefs.model.WidgetSettings

import android.graphics.Bitmap
import android.graphics.Color
import androidx.palette.graphics.Palette

/**
 * @author alex
 * @date 2014-12-06
 */
class CardsBackgroundProcessor : BackgroundProcessor {

    override fun getColor(prefs: WidgetSettings, icon: Bitmap?): Int {
        if (icon == null) {
            return Color.DKGRAY
        }
        val palette = Palette.Builder(icon).generate()

        var swatch: Palette.Swatch? = palette.mutedSwatch

        if (swatch != null) {
            return swatch.rgb
        }
        swatch = palette.darkVibrantSwatch
        if (swatch != null) {
            return swatch.rgb
        }
        swatch = palette.vibrantSwatch
        return swatch?.rgb ?: Color.DKGRAY
    }
}
